pkg/source/postgres.go

```diff
@@ -180,16 +185,21 @@ func (p *PGXSource) fetching(ctx context.Context) (change Change, err error) {
 		switch msg.Data[0] {
 		case pglogrepl.PrimaryKeepaliveMessageByteID:
 			var pkm pglogrepl.PrimaryKeepaliveMessage
-			if pkm, err = pglogrepl.ParsePrimaryKeepaliveMessage(msg.Data[1:]); err == nil && pkm.ReplyRequested {
+			pkm, err = pglogrepl.ParsePrimaryKeepaliveMessage(msg.Data[1:])
+			if err != nil {
+				return change, err
+			}
+
+			c := Change{
+				Checkpoint: cursor.Checkpoint{LSN: uint64(pkm.ServerWALEnd)},
+				Message:    &pb.Message{Type: &pb.Message_KeepAlive{KeepAlive: &pb.KeepAlive{}}},
+			}
+			if pkm.ReplyRequested {
 				p.nextReportTime = time.Time{}
-				// We aim to catch up to the ServerWALEnd position.
-				// For instance, PostgreSQL sends a keepalive message when shutting down.
-				// If we don't report the ServerWALEnd position, the replication connection cannot close.
-				// Consequently, PostgreSQL could be indefinitely stuck in the shutdown process.
-				change = Change{
-					Checkpoint: cursor.Checkpoint{LSN: uint64(pkm.ServerWALEnd)},
-					Message:    &pb.Message{Type: &pb.Message_KeepAlive{KeepAlive: &pb.KeepAlive{}}},
-				}
+				change = c
+			} else if c.Checkpoint.LSN-p.currentLsn >= p.ReportLSNThreshold {
```

Comment:
The ServerWALEnd could also stick to a value if the entire database was idle, resulting in the threshold not be exceeded for an uncertain while. How about using time as threshold?

Reply:
There is another good option. Should we need to use both LSN and time threshold ?